File: curriculum/units/M08/code/python/capability_projection.py
from __future__ import annotations

from dataclasses import dataclass
import re
from types import MappingProxyType
from typing import Callable, Literal, Mapping

CapabilitySource = Literal["builtin", "plugin", "mcp", "dynamic", "policy"]
ProjectionReason = Literal[
    "included",
    "hidden-by-policy",
    "mode-mismatch",
    "provider-mismatch",
    "model-mismatch",
    "deferred-until-discovered",
]


@dataclass(frozen=True)
class CapabilityDefinition:
    name: str
    description: str
    source: CapabilitySource
    priority: float
    deferred: bool = False
    modes: tuple[str, ...] = ()
    providers: tuple[str, ...] = ()
    models: tuple[str, ...] = ()


@dataclass(frozen=True)
class CatalogSnapshot:
    revision: int
    capabilities: tuple[CapabilityDefinition, ...]


class CapabilityCatalog:
    def __init__(self) -> None:
        self._revision = 0
        self._capabilities: dict[str, CapabilityDefinition] = {}

    def publish(
        self, definitions: tuple[CapabilityDefinition, ...] | list[CapabilityDefinition]
    ) -> CatalogSnapshot:
        next_capabilities = dict(self._capabilities)
        for definition in definitions:
            _validate_capability(definition)
            current = next_capabilities.get(definition.name)
            if current is None or definition.priority > current.priority:
                next_capabilities[definition.name] = definition
        self._capabilities = next_capabilities
        self._revision += 1
        return self.snapshot()

    def snapshot(self) -> CatalogSnapshot:
        return CatalogSnapshot(
            revision=self._revision,
            capabilities=tuple(
                sorted(self._capabilities.values(), key=lambda item: item.name)
            ),
        )


@dataclass(frozen=True)
class ProjectionDecision:
    name: str
    source: CapabilitySource
    included: bool
    reason: ProjectionReason


@dataclass(frozen=True)
class ToolSchema:
    name: str
    description: str
    source: CapabilitySource


@dataclass(frozen=True)
class CapabilitySnapshot:
    catalog_revision: int
    boundary: str
    mode: str
    provider: str
    model: str
    schemas: tuple[ToolSchema, ...]
    decisions: tuple[ProjectionDecision, ...]


@dataclass(frozen=True)
class ProjectionOptions:
    boundary: str
    mode: str
    provider: str
    model: str
    policy_hidden_names: frozenset[str] = frozenset()
    discovered_deferred_names: frozenset[str] = frozenset()


class CapabilityProjector:
    def project(
        self, catalog: CatalogSnapshot, options: ProjectionOptions
    ) -> CapabilitySnapshot:
        _require_non_empty(options.boundary, "boundary")
        _require_non_empty(options.mode, "mode")
        _require_non_empty(options.provider, "provider")
        _require_non_empty(options.model, "model")
        decisions = tuple(
            _decide_projection(capability, options)
            for capability in catalog.capabilities
        )
        schemas = tuple(
            ToolSchema(capability.name, capability.description, capability.source)
            for capability, decision in zip(
                catalog.capabilities, decisions, strict=True
            )
            if decision.included
        )
        return CapabilitySnapshot(
            catalog_revision=catalog.revision,
            boundary=options.boundary,
            mode=options.mode,
            provider=options.provider,
            model=options.model,
            schemas=schemas,
            decisions=decisions,
        )


ExecutableHandler = Callable[[object], object]


class ExecutableRegistry:
    def __init__(self) -> None:
        self._handlers: dict[str, ExecutableHandler] = {}

    def register(self, name: str, handler: ExecutableHandler) -> None:
        _require_non_empty(name, "tool name")
        if name in self._handlers:
            raise ValueError(f"executable already registered: {name}")
        self._handlers[name] = handler

    def names(self) -> tuple[str, ...]:
        return tuple(sorted(self._handlers))

    def dispatch(
        self, snapshot: CapabilitySnapshot, name: str, input_value: object
    ) -> object:
        if not any(schema.name == name for schema in snapshot.schemas):
            raise ValueError(
                f"tool is not visible in boundary {snapshot.boundary}: {name}"
            )
        handler = self._handlers.get(name)
        if handler is None:
            raise ValueError(f"visible tool has no executable handler: {name}")
        return handler(input_value)


@dataclass(frozen=True)
class BuiltSystemContext:
    system_prompt: tuple[str, ...]
    meta_user_context: Mapping[str, str]
    system_context: Mapping[str, str]
    base: Literal["default", "custom"]


class SystemContextBuilder:
    def build(
        self,
        *,
        default_prompt: tuple[str, ...] | list[str],
        custom_prompt: str | None = None,
        append_prompt: str | None = None,
        user_context: Mapping[str, str] | None = None,
        system_context: Mapping[str, str] | None = None,
    ) -> BuiltSystemContext:
        custom = custom_prompt.strip() if custom_prompt else ""
        appended = append_prompt.strip() if append_prompt else ""
        base = (custom,) if custom else tuple(default_prompt)
        effective = base + ((appended,) if appended else ())
        if not effective:
            raise ValueError("an effective system prompt is required")
        return BuiltSystemContext(
            system_prompt=effective,
            meta_user_context=MappingProxyType(dict(user_context or {})),
            system_context=MappingProxyType(dict(system_context or {})),
            base="custom" if custom else "default",
        )


def _decide_projection(
    capability: CapabilityDefinition, options: ProjectionOptions
) -> ProjectionDecision:
    reason: ProjectionReason = "included"
    if capability.name in options.policy_hidden_names:
        reason = "hidden-by-policy"
    elif capability.modes and options.mode not in capability.modes:
        reason = "mode-mismatch"
    elif capability.providers and options.provider not in capability.providers:
        reason = "provider-mismatch"
    elif capability.models and options.model not in capability.models:
        reason = "model-mismatch"
    elif (
        capability.deferred
        and capability.name not in options.discovered_deferred_names
    ):
        reason = "deferred-until-discovered"
    return ProjectionDecision(
        name=capability.name,
        source=capability.source,
        included=reason == "included",
        reason=reason,
    )


def _validate_capability(definition: CapabilityDefinition) -> None:
    _require_non_empty(definition.name, "capability name")
    if re.fullmatch(r"[A-Za-z][A-Za-z0-9_.:-]*", definition.name) is None:
        raise ValueError(
            "capability name must use the portable ASCII identifier form"
        )
    _require_non_empty(definition.description, "capability description")
    if definition.priority != definition.priority or definition.priority in (
        float("inf"),
        float("-inf"),
    ):
        raise ValueError("capability priority must be finite")


def _require_non_empty(value: str, name: str) -> None:
    if not value.strip():
        raise ValueError(f"{name} must not be empty")

File: curriculum/units/M08/code/typescript/capability-projection.ts
export type CapabilitySource =
  | 'builtin'
  | 'plugin'
  | 'mcp'
  | 'dynamic'
  | 'policy'

export type CapabilityDefinition = Readonly<{
  name: string
  description: string
  source: CapabilitySource
  priority: number
  deferred?: boolean
  modes?: readonly string[]
  providers?: readonly string[]
  models?: readonly string[]
}>

export type CatalogSnapshot = Readonly<{
  revision: number
  capabilities: readonly CapabilityDefinition[]
}>

export class CapabilityCatalog {
  #revision = 0
  #capabilities = new Map<string, CapabilityDefinition>()

  publish(definitions: readonly CapabilityDefinition[]): CatalogSnapshot {
    const next = new Map(this.#capabilities)
    for (const definition of definitions) {
      validateCapability(definition)
      const candidate = freezeCapability(definition)
      const current = next.get(candidate.name)
      if (!current || candidate.priority > current.priority) {
        next.set(candidate.name, candidate)
      }
    }
    this.#capabilities = next
    this.#revision += 1
    return this.snapshot()
  }

  snapshot(): CatalogSnapshot {
    const capabilities = [...this.#capabilities.values()]
      .sort((left, right) => compareAsciiNames(left.name, right.name))
      .map(freezeCapability)
    return Object.freeze({
      revision: this.#revision,
      capabilities: Object.freeze(capabilities),
    })
  }
}

export type ProjectionReason =
  | 'included'
  | 'hidden-by-policy'
  | 'mode-mismatch'
  | 'provider-mismatch'
  | 'model-mismatch'
  | 'deferred-until-discovered'

export type ProjectionDecision = Readonly<{
  name: string
  source: CapabilitySource
  included: boolean
  reason: ProjectionReason
}>

export type ToolSchema = Readonly<{
  name: string
  description: string
  source: CapabilitySource
}>

export type CapabilitySnapshot = Readonly<{
  catalogRevision: number
  boundary: string
  mode: string
  provider: string
  model: string
  schemas: readonly ToolSchema[]
  decisions: readonly ProjectionDecision[]
}>

export type ProjectionOptions = Readonly<{
  boundary: string
  mode: string
  provider: string
  model: string
  policyHiddenNames?: ReadonlySet<string>
  discoveredDeferredNames?: ReadonlySet<string>
}>

export class CapabilityProjector {
  project(
    catalog: CatalogSnapshot,
    options: ProjectionOptions,
  ): CapabilitySnapshot {
    requireNonEmpty(options.boundary, 'boundary')
    requireNonEmpty(options.mode, 'mode')
    requireNonEmpty(options.provider, 'provider')
    requireNonEmpty(options.model, 'model')

    const decisions = catalog.capabilities.map(capability =>
      decideProjection(capability, options),
    )
    const schemas = catalog.capabilities
      .filter((_, index) => decisions[index]?.included)
      .map(capability =>
        Object.freeze({
          name: capability.name,
          description: capability.description,
          source: capability.source,
        }),
      )

    return Object.freeze({
      catalogRevision: catalog.revision,
      boundary: options.boundary,
      mode: options.mode,
      provider: options.provider,
      model: options.model,
      schemas: Object.freeze(schemas),
      decisions: Object.freeze(decisions),
    })
  }
}

export type ExecutableHandler = (input: unknown) => unknown

export class ExecutableRegistry {
  #handlers = new Map<string, ExecutableHandler>()

  register(name: string, handler: ExecutableHandler): void {
    requireNonEmpty(name, 'tool name')
    if (this.#handlers.has(name)) {
      throw new Error(`executable already registered: ${name}`)
    }
    this.#handlers.set(name, handler)
  }

  names(): readonly string[] {
    return Object.freeze([...this.#handlers.keys()].sort(compareAsciiNames))
  }

  dispatch(snapshot: CapabilitySnapshot, name: string, input: unknown): unknown {
    if (!snapshot.schemas.some(schema => schema.name === name)) {
      throw new Error(`tool is not visible in boundary ${snapshot.boundary}: ${name}`)
    }
    const handler = this.#handlers.get(name)
    if (!handler) {
      throw new Error(`visible tool has no executable handler: ${name}`)
    }
    return handler(input)
  }
}

export type BuiltSystemContext = Readonly<{
  systemPrompt: readonly string[]
  metaUserContext: Readonly<Record<string, string>>
  systemContext: Readonly<Record<string, string>>
  base: 'default' | 'custom'
}>

export class SystemContextBuilder {
  build(input: {
    defaultPrompt: readonly string[]
    customPrompt?: string
    appendPrompt?: string
    userContext?: Readonly<Record<string, string>>
    systemContext?: Readonly<Record<string, string>>
  }): BuiltSystemContext {
    const customPrompt = input.customPrompt?.trim()
    const appendPrompt = input.appendPrompt?.trim()
    const base = customPrompt ? [customPrompt] : [...input.defaultPrompt]
    const systemPrompt = appendPrompt ? [...base, appendPrompt] : base
    if (systemPrompt.length === 0) {
      throw new Error('an effective system prompt is required')
    }
    return Object.freeze({
      systemPrompt: Object.freeze(systemPrompt),
      metaUserContext: Object.freeze({ ...(input.userContext ?? {}) }),
      systemContext: Object.freeze({ ...(input.systemContext ?? {}) }),
      base: customPrompt ? 'custom' : 'default',
    })
  }
}

function decideProjection(
  capability: CapabilityDefinition,
  options: ProjectionOptions,
): ProjectionDecision {
  let reason: ProjectionReason = 'included'
  if (options.policyHiddenNames?.has(capability.name)) {
    reason = 'hidden-by-policy'
  } else if (capability.modes && !capability.modes.includes(options.mode)) {
    reason = 'mode-mismatch'
  } else if (
    capability.providers &&
    !capability.providers.includes(options.provider)
  ) {
    reason = 'provider-mismatch'
  } else if (capability.models && !capability.models.includes(options.model)) {
    reason = 'model-mismatch'
  } else if (
    capability.deferred &&
    !options.discoveredDeferredNames?.has(capability.name)
  ) {
    reason = 'deferred-until-discovered'
  }
  return Object.freeze({
    name: capability.name,
    source: capability.source,
    included: reason === 'included',
    reason,
  })
}

function freezeCapability(
  definition: CapabilityDefinition,
): CapabilityDefinition {
  return Object.freeze({
    ...definition,
    ...(definition.modes ? { modes: Object.freeze([...definition.modes]) } : {}),
    ...(definition.providers
      ? { providers: Object.freeze([...definition.providers]) }
      : {}),
    ...(definition.models
      ? { models: Object.freeze([...definition.models]) }
      : {}),
  })
}

function validateCapability(definition: CapabilityDefinition): void {
  requireNonEmpty(definition.name, 'capability name')
  if (!/^[A-Za-z][A-Za-z0-9_.:-]*$/.test(definition.name)) {
    throw new Error('capability name must use the portable ASCII identifier form')
  }
  requireNonEmpty(definition.description, 'capability description')
  if (!Number.isFinite(definition.priority)) {
    throw new Error('capability priority must be finite')
  }
}

function requireNonEmpty(value: string, name: string): void {
  if (value.trim() === '') throw new Error(`${name} must not be empty`)
}

function compareAsciiNames(left: string, right: string): number {
  return left < right ? -1 : left > right ? 1 : 0
}
